File: CountNumberOfNiceSubarrays.cpp
class Solution {
public:
    int numberOfSubarrays(vector<int>& nums, int k) {
        
        int n = nums.size();
        unordered_map<int, int> prefixSum;
        int sum = (nums[0] % 2 == 0)?0:1;
        prefixSum[sum] += 1;
        int answer = 0;

        if (sum == k)
            answer++;

        for (int i = 1; i < n; i++)
        {
            sum += (nums[i] % 2 == 0)?0:1;
            if (prefixSum.find(sum - k) != prefixSum.end())
                answer += prefixSum[sum - k];

                    prefixSum[sum] += 1;

            if (sum == k)
            {
                answer++;
                continue;
            }
        }

        return answer;
        
    }
};

File: CountNumberOfNiceSubarrays.py
class Solution:
    def numberOfSubarrays(self, nums: List[int], k: int) -> int:
        n = len(nums)
        left = 0
        right = 0
        evenAfterK = 0
        count = 0
        nice = 0
        nums.append(0)

        for right in range(n+1):

            if nums[right] % 2 != 0:
                count += 1

            while (count == k and right == n) or count > k:
                nice += evenAfterK
                if nums[left] % 2 != 0:
                    count -= 1
                left += 1
                if not ((count == k and right == n) or count > k):
                    evenAfterK = 0

            
            if count == k:
                evenAfterK += 1


        return nice
